crates/sail-python-udf/src/udf/pyspark_udf.rs

```diff
@@ -79,175 +253,39 @@ impl ScalarUDFImpl for PySparkUDF {
-                Ok(array_data)
-            })?;
-            return Ok(ColumnarValue::Array(make_array(array_data)));
-        }
-
-        let array_data = Python::with_gil(|py| -> PyUdfResult<_> {
-            let pyarrow_module_array = get_pyarrow_array_function(py)?;
-            let builtins_list = get_python_builtins_list_function(py)?;
-            let builtins_str = get_python_builtins_str_function(py)?;
-            let python_function = self.python_function.function(py)?;
-            let pyarrow_output_data_type = get_pyarrow_output_data_type(&self.output_type, py)?;
-            let pyarrow_array_kwargs =
-                build_pyarrow_array_kwargs(py, pyarrow_output_data_type, false)?;
-
-            let py_args_columns_list = args
-                .iter()
-                .map(|arg| {
-                    let arg = arg
-                        .into_data()
-                        .to_pyarrow(py)?
-                        .call_method0(py, intern!(py, "to_pylist"))?
-                        .clone_ref(py)
-                        .into_bound(py);
-                    Ok(arg)
-                })
-                .collect::<PyUdfResult<Vec<_>>>()?;
-            let py_args_tuple = PyTuple::new_bound(py, &py_args_columns_list);
-            // TODO: Do zip in Rust for performance.
-            let py_args_zip = py.eval_bound("zip", None, None)?.call1(&py_args_tuple)?;
-            let py_args = PyIterator::from_bound_object(&py_args_zip)?;
-
-            let mut already_str: bool = false;
-            let results = py_args
-                .map(|py_arg| -> PyUdfResult<Bound<PyAny>> {
-                    let py_arg = py_arg?;
-                    let result = python_function.call1((py.None(), (py_arg,)))?;
-                    let result = builtins_list.call1((result,))?.get_item(0)?;
-
-                    if matches!(self.eval_type, spec::PySparkUdfType::Batched)
-                        && self.deterministic
-                        && self.output_type == DataType::Utf8
-                    {
-                        if already_str {
-                            return Ok(result);
-                        }
-                        let result_type = result.get_type();
-                        let result_data_type_name = result_type.name()?;
-                        if result_data_type_name != "str" {
-                            let result = builtins_str.call1((result,))?;
-                            return Ok(result);
-                        } else {
-                            already_str = true;
-                        }
-                    }
-                    Ok(result)
-                })
-                .collect::<Result<Vec<_>, _>>()?;
-
-            let results = pyarrow_module_array.call((results,), Some(&pyarrow_array_kwargs))?;
-
-            Ok(ArrayData::from_pyarrow_bound(&results)?)
-        })?;
-
+        let array_data = if self.eval_type.is_arrow_udf() {
```

Comment:
After this small refactoring, the duplicated code becomes more obvious. We can clean up them in future PRs.